alf/algorithms/algorithm.py

```diff
@@ -206,124 +169,95 @@ def _trainable_attributes_to_ignore(self):
-            new_var_ids = set(map(id, new_vars))
-            dup_ids = var_ids & new_var_ids
-            assert not dup_ids, (
-                "Modules/variables %s might have multiple optimizers! Consider "
-                "specifying attributes in _trainable_attributes_to_ignore()" %
-                new_module_or_var.name)
-            var_ids.update(new_var_ids)
-
-        for alg in self._get_children(_is_alg):
-            opt_and_module = alg.get_optimizer_and_module_sets()
-            for alg_name, opt, module_set in opt_and_module:
-                for m in module_set:
-                    _check_module_or_var(m)
-                if opt is not None:
-                    optimizers.append(opt)
-                    algorithm_names.append(alg_name)
-                    module_sets.append(module_set)
-                else:
-                    module_sets[0].extend(module_set)
+        for name, param in self.named_parameters(recurse=False):
+            if name in self._trainable_attributes_to_ignore():
+                continue
+            children.append(param)
 
-        for var in self._get_children(_is_trainable_var):
-            if id(var) not in init_module_ids:  # exclude already init-ed
-                _check_module_or_var(var)
-                module_sets[0].append(var)
+        return children
 
-        # Prefer to report errors on module level
-        for module in self._get_children(_is_trainable_module):
-            if id(module) not in init_module_ids:  # exclude already init-ed
-                _check_module_or_var(module)
-                module_sets[0].append(module)
+    @property
+    def default_optimizer(self):
+        """Get the default optimizer for this algorithm."""
+        return self._default_optimizer
 
-        return list(zip(algorithm_names, optimizers, module_sets))
+    def _setup_optimizers(self):
+        """Setup the param groups for optimizers.
 
-    def _get_children(self, predicate):
-        module_dict = vars(self)
-        children = []
-        ids = set()
-        for key in sorted(module_dict):
-            attr = module_dict[key]
-            if (key in [
-                    '_init_optimizers', '_cached_opt_and_var_sets',
-                    '_init_module_sets'
-            ] or key in self._TF_MODULE_IGNORED_PROPERTIES
-                    or key in self._trainable_attributes_to_ignore()):
+        Returns:
+            list of parameters not handled by any optimizers under this algorithm
+        """
+        default_optimizer = self.default_optimizer
+        new_params = []
+        handled = set()
+        for child in self._get_children():
+            if child in handled:
                 continue
-            for leaf in tf.nest.flatten(attr):
-                if predicate(leaf) and not id(leaf) in ids:
-                    children.append(leaf)
-                    ids.add(id(leaf))
-        return children
+            handled.add(child)
```

Comment:
Is it possible to detect a loop and report errors if a grandchild contains the current algorithm as a child?